command/job_restart.go

```diff
@@ -0,0 +1,850 @@
+
+  -verbose
+    Display full information.
+`
+	return strings.TrimSpace(helpText)
+}
+
+func (c *JobRestartCommand) Synopsis() string {
+	return "Restart or reschedule allocations for a job"
+}
+
+func (c *JobRestartCommand) AutocompleteFlags() complete.Flags {
+	return mergeAutocompleteFlags(c.Meta.AutocompleteFlags(FlagSetClient),
+		complete.Flags{
+			"-all-tasks":         complete.PredictNothing,
+			"-batch-size":        complete.PredictAnything,
+			"-batch-wait":        complete.PredictAnything,
+			"-fail-on-error":     complete.PredictNothing,
+			"-no-shutdown-delay": complete.PredictNothing,
+			"-reschedule":        complete.PredictNothing,
+			"-task":              complete.PredictAnything,
+			"-verbose":           complete.PredictNothing,
+		})
+}
+
+func (c *JobRestartCommand) AutocompleteArgs() complete.Predictor {
+	return complete.PredictFunc(func(a complete.Args) []string {
+		client, err := c.Meta.Client()
+		if err != nil {
+			return nil
+		}
+
+		resp, _, err := client.Search().PrefixSearch(a.Last, contexts.Jobs, nil)
+		if err != nil {
+			return []string{}
+		}
+		return resp.Matches[contexts.Jobs]
+	})
+}
+
+func (c *JobRestartCommand) Name() string { return "job restart" }
+
+func (c *JobRestartCommand) Run(args []string) int {
+	c.ui = &cli.ConcurrentUi{Ui: c.Ui}
+
+	// Handle SIGINT to prevent accidental cancellations.
+	// activeCh is blocked while a signal is being handled.
+	activeCh := make(chan interface{})
+	sigsCh := make(chan os.Signal, 1)
+	signal.Notify(sigsCh, syscall.SIGINT)
+	go c.handleSignal(sigsCh, activeCh)
+
+	// Parse and validate command line arguments.
+	err, code := c.parseAndValidate(args)
+	if err != nil {
+		c.ui.Error(err.Error())
+		c.ui.Error(commandErrorText(c))
+		return code
+	}
+	if code != 0 {
+		return code
+	}
+
+	c.client, err = c.Meta.Client()
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("Error initializing client: %v", err))
+		return 1
+	}
+
+	// Check if the job exists.
+	// Avoid prefix matching to make sure we only restart the expected job.
+	job, _, err := c.client.Jobs().Info(c.jobID, nil)
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("Error retrieving job %q: %s", c.jobID, err))
+		return 1
+	}
+
+	// Check if passed groups and tasks are valid.
+	err = c.validateGroupsAndTasks(job)
+	if err != nil {
+		c.ui.Error(err.Error())
+		return 1
+	}
+
+	// Fetch all allocations for the job and filter out the ones that are not
+	// eligible for restart.
+	allocs, _, err := c.client.Jobs().Allocations(c.jobID, false, nil)
+	if err != nil {
+		c.ui.Error(fmt.Sprintf("Error retrieving allocations for jobs %s: %v", c.jobID, err))
+		return 1
+	}
+
+	restartAllocs := make([]string, 0, len(allocs))
+	for _, alloc := range allocs {
+		shortAllocID := limit(alloc.ID, c.length)
+
+		// Skip allocations that are not running.
+		allocRunning := alloc.ClientStatus == api.AllocClientStatusRunning && alloc.DesiredStatus == api.AllocDesiredStatusRun
+		if !allocRunning {
```

Comment:
Hm, should this be an `||` instead of an `&&` so the command works on allocs that **either** running or desired to be run?

For example:
1. If an alloc has `Client=Running, Desired=Stop`, it may be in the user still may want `-no-shutdown-delay` to kick the alloc immediately.
2. If an alloc has `Client=Pending, Desired=Run`, users may still want to reschedule it off a problematic client (that is failing to start it maybe?)

These are pretty narrow edge cases, so I'm not too worried about it. However, I think it's pretty safe to be aggressive here and try to kick more things instead of fewer things. The scheduler should never run too many allocations or resurrect dead allocs.

Reply:
Ah good point. I first thought it wouldn't make sense to restart an alloc that is already marked to stop anyway, but it could be stuck or something else. I agree that changing this seems fine 👍